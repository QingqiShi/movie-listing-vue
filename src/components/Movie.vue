<template>
  <div class="movie">
    <movie-image
      class="movie__image"
      :src="movie.poster_path"
      type="poster"
      :target-width="400"
    />
    <transition-group tag="div" class="movie__badge" name="flip-list">
      <div
        v-for="badge in badgeList"
        :key="`movie-badge-${movie.id}-${badge.name}`"
        :class="badge.class"
      >
        <icon :name="badge.icon" />
        <span>{{ badge.value }}</span>
      </div>
    </transition-group>
    <div class="movie__genres">
      <div
        v-for="genreId in movie.genre_ids"
        :key="`movie-genres-${movie.id}-${genreId}`"
        class="movie__genres-genre"
      >
        <chip>{{ genres.find(genre => genre.id === genreId).name }}</chip>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState } from "vuex";
import Icon from "@/components/Icon.vue";
import MovieImage from "@/components/MovieImage.vue";
import Chip from "@/components/Chip.vue";

export default {
  components: { Icon, MovieImage, Chip },
  props: {
    movie: {
      type: Object,
      required: true
    }
  },
  computed: {
    ...mapState({
      genres: state => state.movies.genres
    }),
    sort() {
      return this.$store.state.movies.sortProperty;
    },
    badgeList() {
      const list = [
        {
          name: "popularity",
          class: {
            movie__popularity: true,
            "movie__badge--active": this.sort === "popularity"
          },
          property: "popularity",
          icon: "whatshot",
          value: this.movie.popularity.toFixed(1)
        },
        {
          name: "rating",
          class: {
            movie__rating: true,
            "movie__badge--active": this.sort === "vote_average"
          },
          property: "vote_average",
          icon: "star",
          value: this.movie.vote_average
        }
      ];

      return this.sort === "popularity" ? list : list.reverse();
    }
  }
};
</script>

<style lang="scss">
.movie {
  height: 0;
  padding-bottom: 150%;
  position: relative;
  overflow: hidden;

  &:focus {
    outline: none;
  }

  &__image,
  &__badge {
    will-change: transform;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }

  &__image {
    transition: 0.2s transform;

    .movie:hover & {
      transform: scale(1.2);
    }
  }

  &__popularity,
  &__rating {
    display: flex;
    padding-bottom: 0.3rem;

    .material-icons {
      vertical-align: sub;
      display: inline;
      margin-right: 0.5rem;
      line-height: 1rem;
      font-size: 0.8rem;
    }

    span {
      font-family: $font-heading;
      display: block;
      flex-grow: 1;
      line-height: 1rem;
      font-size: 0.8rem;
    }
  }

  &__badge {
    transition: 0.2s font-size;
    padding: 0.5rem;
    background: linear-gradient(
      135deg,
      RGBA($primary-rgb, 1) 0%,
      RGBA($primary-rgb, 0) 15%
    );

    .movie:hover & {
      background: linear-gradient(
        135deg,
        RGBA($accent-rgb, 1) 0%,
        RGBA($accent-rgb, 0) 15%
      );
    }

    &--active {
      span,
      .material-icons {
        font-size: 1.2rem;
      }
    }
  }

  &__genres {
    position: absolute;
    bottom: 0;
    left: 0;
    right: 0;
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    padding: 0 2rem 2rem;
    overflow: auto;
    opacity: 0;
    max-height: 100%;
    transition: 0.2s opacity;

    &::-webkit-scrollbar {
      display: none;
    }

    .movie:hover & {
      opacity: 1;
    }

    &-genre {
      margin-bottom: 0.5rem;
    }
  }
}
</style>
