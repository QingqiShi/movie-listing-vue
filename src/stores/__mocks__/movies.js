const state = {
  configuration: {
    imageBaseUrl: "https://images.com/",
    posterSizes: ["w100", "w300", "original"],
    backdropSizes: ["w100", "w300", "original"]
  },
  movies: [
    {
      id: 1000,
      title: "Avatar",
      genre_ids: [0],
      popularity: 230,
      vote_average: 5
    }
  ],
  genres: [{ id: 0, name: "Action" }, { id: 1, name: "Horror" }],
  sortOrder: false,
  sortProperty: "popularity",
  filterRating: 3,
  filterGenres: { 0: false, 1: true },
  error: ""
};

const getters = {
  /** @param {typeof state} state */
  ratingFilteredMovies: state => state.movies,
  availableGenres: () => [{ id: 0, name: "Action" }],
  filteredMovies: () => []
};

const mutations = {
  setConfiguration: jest.fn(),
  setMovies: jest.fn(),
  setGenres: jest.fn(),
  setSort: jest.fn(),
  setFilter: jest.fn(),
  setError: jest.fn()
};

const actions = {
  initialFetch: jest.fn(() => Promise.resolve()),
  fetchConfiguration: jest.fn(() => Promise.resolve()),
  fetchMovies: jest.fn(() => Promise.resolve()),
  fetchGenres: jest.fn(() => Promise.resolve())
};

export default { state, getters, mutations, actions };
