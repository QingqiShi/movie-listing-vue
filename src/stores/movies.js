import api from "@/constants/api";
import app from "@/constants/app";

const { CONFIGURATION, NOW_PLAYING, GENRES_LIST } = api;
const { LOCALE } = app;

const state = {
  configuration: {},
  /** @type {Movie[]} List of all movies */
  movies: [],
  /** @type {Genre[]} List of all genres */
  genres: [],
  /** @type {boolean} True for accending, false for decending */
  sortOrder: false,
  /** @type {"popularity"|"vote_average"} */
  sortProperty: "popularity",
  /** @type {number} Number between 0 and 10 */
  filterRating: 3,
  /** @type {{[id:number]: boolean}} */
  filterGenres: {},
  /** @type {string} */
  error: ""
};

const getters = {
  /** @type {(state:State) => Movie[]} */
  ratingFilteredMovies: state =>
    state.movies.filter(movie => movie.vote_average >= state.filterRating),

  /** @type {(state:State, getters:Getters) => Genre[]} */
  availableGenres: (state, getters) =>
    state.genres.filter(genre => {
      for (let movie of getters.ratingFilteredMovies) {
        if (movie.genre_ids.includes(genre.id)) {
          return true;
        }
      }
      return state.filterGenres[genre.id];
    }),

  /** @type {(state:State, getters:Getters) => Movie[]} */
  filteredMovies: (state, getters) => {
    const { availableGenres, ratingFilteredMovies } = getters;
    const { filterGenres } = state;

    const genres = availableGenres.filter(genre => filterGenres[genre.id]);
    if (!genres.length) return ratingFilteredMovies;

    return ratingFilteredMovies.filter(({ genre_ids }) =>
      genres.reduce((show, genre) => show && genre_ids.includes(genre.id), true)
    );
  }
};

const mutations = {
  /** @type {(state:State, payload:Configuration) => void} */
  setConfiguration(state, { imageBaseUrl, posterSizes, backdropSizes }) {
    state.configuration = { imageBaseUrl, posterSizes, backdropSizes };
  },

  /** @type {(state:State, movies:Movie[]) => void} */
  setMovies(state, movies) {
    state.movies = movies;
  },

  /** @type {(state:State, genres:Genre[]) => void} */
  setGenres(state, genres) {
    state.genres = genres;
    state.filterGenres = {};
    genres.forEach(genre => {
      state.filterGenres[genre.id] = false;
    });
  },

  /** @type {(state:State, payload:{property:SortProperty, order:boolean}) => void} */
  setSort(state, { property, order }) {
    state.sortOrder = order;
    state.sortProperty = property;
    state.movies = state.movies.sort((a, b) =>
      order ? a[property] - b[property] : b[property] - a[property]
    );
  },

  /** @type {(state:State, payload:{type:"rating"|"genres", value:number}) => void} */
  setFilter(state, { type, value }) {
    if (type === "rating") {
      state.filterRating = value;
    } else {
      const newFilters = { ...state.filterGenres };
      newFilters[value] = !state.filterGenres[value];
      state.filterGenres = newFilters;
    }
  },

  /** @type {(state:State, errorMsg:string) => void} */
  setError(state, errorMsg) {
    state.error = errorMsg;
  }
};

const actions = {
  /** @param {import('vuex').ActionContext} context */
  async initialFetch({ commit, dispatch }) {
    try {
      await Promise.all([
        dispatch("fetchConfiguration"),
        dispatch("fetchMovies"),
        dispatch("fetchGenres")
      ]);
      commit("setSort", { property: "popularity", order: false });
    } catch (e) {
      commit("setError", e);
    }
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchConfiguration({ commit }) {
    const result = await fetchApi(
      CONFIGURATION(),
      "FetchConfiguration",
      result => result && result.images
    );

    commit("setConfiguration", {
      imageBaseUrl: result.images.base_url,
      posterSizes: result.images.poster_sizes,
      backdropSizes: result.images.backdrop_sizes
    });
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchMovies({ commit }) {
    const result = await fetchApi(
      NOW_PLAYING(LOCALE),
      "FetchMovies",
      result => result && result.results
    );

    commit("setMovies", result.results);
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchGenres({ commit }) {
    const result = await fetchApi(
      GENRES_LIST(LOCALE),
      "FetchGenres",
      result => result && result.genres
    );

    commit("setGenres", result.genres);
  }
};

/**
 * @param {string} url
 * @param {string} name
 */
async function fetchApi(url, name, validator) {
  let result;

  try {
    const response = await fetch(url);
    result = await response.json();
  } catch {
    throw `${name} failed, check your network and try again`;
  }

  if (!validator(result)) {
    throw `${name} returned invalid result`;
  }

  return result;
}

export default { state, getters, mutations, actions };

/* Type definitions to help vscode do auto suggest */

/** @typedef {{id:number, name:string}} Genre */
/**
 * @typedef {object} Movie
 * @property {number} Movie.id
 * @property {string} Movie.title
 * @property {number} Movie.popularity
 * @property {number} Movie.vote_average
 * @property {number[]} Movie.genre_ids
 * @property {string} Movie.poster_path
 */
/**
 * @typedef {object} Configuration
 * @property {string} Configuration.imageBaseUrl
 * @property {string} Configuration.posterSizes
 * @property {string} Configuration.backdropSizes
 */
/** @typedef {typeof state} State */
/**
 * @typedef {object} Getters
 * @property {Movie[]} Getters.ratingFilteredMovies
 * @property {Genre[]} Getters.availableGenres
 * @property {Movie[]} Getters.filteredMovies
 */
/** @typedef {typeof state.sortProperty} SortProperty */
