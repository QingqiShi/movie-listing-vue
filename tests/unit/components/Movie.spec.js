import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Movie from "@/components/Movie.vue";
import movies from "@/stores/movies";

jest.mock("@/stores/movies");

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Movie.vue", () => {
  let wrapper;

  beforeAll(() => {
    const store = new Vuex.Store({ modules: { movies } });
    wrapper = shallowMount(Movie, {
      store,
      localVue,
      propsData: { movie: movies.state.movies[0] }
    });
  });

  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("reverse badge based on sort property", () => {
    wrapper.vm.$store.state.movies.sortProperty = "vote_average";
    expect(wrapper).toMatchSnapshot();
  });
});
