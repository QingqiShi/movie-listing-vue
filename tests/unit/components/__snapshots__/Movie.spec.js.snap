// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie.vue renders 1`] = `
<div class="movie">
  <movie-image-stub src="" type="poster" targetwidth="400" class="movie__image"></movie-image-stub> <span tag="div" name="flip-list" class="movie__badge"><div class="movie__popularity movie__badge--active"><icon-stub name="whatshot"></icon-stub> <span>230.0</span>
</div>
<div class="movie__rating">
  <icon-stub name="star"></icon-stub> <span>5</span>
</div></span>
<div class="movie__genres">
  <div class="movie__genres-genre">
    <chip-stub>Action</chip-stub>
  </div>
</div>
</div>
`;

exports[`Movie.vue reverse badge based on sort property 1`] = `
<div class="movie">
  <movie-image-stub src="" type="poster" targetwidth="400" class="movie__image"></movie-image-stub> <span tag="div" name="flip-list" class="movie__badge"><div class="movie__rating movie__badge--active"><icon-stub name="star"></icon-stub> <span>5</span>
</div>
<div class="movie__popularity">
  <icon-stub name="whatshot"></icon-stub> <span>230.0</span>
</div></span>
<div class="movie__genres">
  <div class="movie__genres-genre">
    <chip-stub>Action</chip-stub>
  </div>
</div>
</div>
`;
